jetty-client/src/main/scala/org/http4s/client/jetty/JettyClient.scala

```diff
@@ -0,0 +1,144 @@
+            try {
+              client.stop()
+            } catch {
+              case NonFatal(t) => logger.error(t)("Unable to shut down Jetty client")
+            }
+          })
+        )
+      })
+
+  def stream[F[_]](client: HttpClient = new HttpClient())(
+      implicit F: ConcurrentEffect[F],
+      ec: ExecutionContext): Stream[F, Client[F]] =
+    Stream.bracket(apply(client))(_.shutdown)
+
+  private def toRequest[F[_]](client: HttpClient, request: Request[F])(
+      implicit F: Effect[F],
+      ec: ExecutionContext): JettyRequest = {
+    val jReq = client.newRequest(request.uri.toString).method(request.method.name)
+
+    for (h <- request.headers)
+      jReq.header(h.name.toString, h.value)
+
+    jReq.content(requestContentProvider(request))
+  }
+
+  private def requestContentProvider[F[_]](
+      request: Request[F])(implicit F: Effect[F], ec: ExecutionContext) = {
+    val dcp = new DeferredContentProvider()
+
+    ec.execute(new Runnable {
+      override def run(): Unit =
+        request.body.chunks
+          .map(_.toByteBuffer)
+          .to(Sink((b: ByteBuffer) => F.delay { dcp.offer(b); () }))
+          .compile
+          .drain
+          .runAsync {
+            case Left(t) =>
+              IO { logger.error(t)("Unable to write to Jetty sink") }
+            case Right(_) =>
+              IO.unit
+          }
+          .flatTap(_ => IO { dcp.close() })
+          .unsafeRunSync()
+    })
+
+    dcp
+  }
+
+  private def responseHandler[F[_]](cb: Callback[DisposableResponse[F]])(
+      implicit F: ConcurrentEffect[F],
+      ec: ExecutionContext): JettyResponse.Listener =
+    async
+      .unboundedQueue[F, Option[ByteBuffer]]
+      .map(queue => {
+        new JettyResponse.Listener.Adapter {
+          override def onHeaders(response: JettyResponse): Unit = {
+            val status = Status.fromInt(response.getStatus)
+
+            val dr = status
+              .map { s =>
+                DisposableResponse[F](
+                  response = Response(
+                    status = s,
+                    headers = getHeaders(response.getHeaders),
+                    body =
+                      queue.dequeue.unNoneTerminate.flatMap(bBuf => chunk(Chunk.byteBuffer(bBuf)))
+                  ),
+                  F.unit
+                )
+              }
+              .leftMap { t =>
+                //this aborts both the response and the request
+                if (!response.abort(t))
+                  IO { logger.error("Failed to abort the response") }
+                t
+              }
+
+            ec.execute(new Runnable {
+              override def run(): Unit = cb(dr)
+            })
+          }
+
+          override def onContent(response: JettyResponse, content: ByteBuffer): Unit = {
+            val copy = ByteBuffer.allocate(content.remaining())
+            copy.put(content).flip()
+            queue.enqueue1(copy.some).runAsync(loggingAsyncCallback(logger)).unsafeRunSync()
```

Comment:
I guess we lack of some kind of back pressure mechanism here too @rossabaker ; we could use the `onContent(Response response, ByteBuffer content, Callback callback)` method here and invoke the `callback` to signal that we need more content (the `callback.succeed()` doesn't control only the "demand content" semantic but also the releasing of the ByteBuffer, see [here](https://github.com/eclipse/jetty.project/issues/2429), I'm not sure how big of a problem it would be here)

Reply:
Ooh, yeah, good point.  The queue can be a bounded queue, but we don't want to block of spin.  Yeah, I think something with the callback would be the right approach here, too.